crates/red_knot_python_semantic/resources/mdtest/snapshots/unsupported_bool_conversion.md_-_Different_ways_that_`UNSUPPORTED_BOOL_CONVERSION`_can_occur_-_Part_of_a_union_where_at_least_one_member_has_incorrect_`__bool__`_method.snap

```diff
@@ -0,0 +1,43 @@
+---
+source: crates/red_knot_test/src/lib.rs
+expression: snapshot
+---
+---
+mdtest name: unsupported_bool_conversion.md - Different ways that `UNSUPPORTED_BOOL_CONVERSION` can occur - Part of a union where at least one member has incorrect `__bool__` method
+mdtest path: crates/red_knot_python_semantic/resources/mdtest/diagnostics/unsupported_bool_conversion.md
+---
+
+# Python source files
+
+## mdtest_snippet.py
+
+```
+ 1 | class NotBoolable1:
+ 2 |     def __bool__(self) -> str:
+ 3 |         return "wat"
+ 4 | 
+ 5 | class NotBoolable2:
+ 6 |     pass
+ 7 | 
+ 8 | class NotBoolable3:
+ 9 |     __bool__: int = 3
+10 | 
+11 | def get() -> NotBoolable1 | NotBoolable2 | NotBoolable3:
+12 |     return NotBoolable2()
+13 | 
+14 | # error: [unsupported-bool-conversion]
+15 | 10 and get() and True
+```
+
+# Diagnostics
+
+```
+error: lint:unsupported-bool-conversion: Boolean conversion is unsupported for union `NotBoolable1 | NotBoolable2 | NotBoolable3` because `NotBoolable1` doesn't implement `__bool__` correctly
```

Comment:
In this case I think ideally we would provide the same information about _why_ `NotBoolable1` doesn't implement `__bool__` correctly as we do in the non-union examples above.

Reply:
Yeah, 100%. I looked into doing this too, but seemed like more work that I wanted to bite off here. I even specifically wrote this test such that `NotBoolable1` and `NotBoolable3` are both not usable as booleans for different reasons, thinking that I could improve that case. And perhaps even include diagnostics for _all_ (or at least more than 1) parts of a union. There's some refactoring involved so that we re-use the diagnostic generation for each of the base cases, and then some way to figure out how to tie off the recursion of union types themselves.